third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb

```diff
@@ -0,0 +1,187 @@
+			},
+
+			"on_delete": {
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringInSlice([]string{"cancel", "drain"}, false),
+				Optional:     true,
+				Default:      "drain",
+			},
+
+			"parameters": {
+				Type:     schema.TypeMap,
+				ForceNew: true,
+				Optional: true,
+			},
+
+			"project": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				// ForceNew applies to both stream and batch jobs
+				ForceNew: true,
+			},
+
+			"state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+// resourceDataflowFlexTemplateJobCreate creates a Flex Template Job from TF code.
+func resourceDataflowFlexTemplateJobCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region, err := getRegion(d, config)
+	if err != nil {
+		return err
+	}
+
+	request := buildLaunchFlexTemplateRequest(d, config)
+
+	job, err := resourceDataflowFlexTemplateJobLaunchTemplate(config, project, region, &request)
+	if err != nil {
+		return err
+	}
+	d.SetId(job.Id)
+
+	return resourceDataflowFlexTemplateJobRead(d, meta)
+}
+
+// resourceDataflowFlexTemplateJobRead reads a Flex Template Job resource.
+func resourceDataflowFlexTemplateJobRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region, err := getRegion(d, config)
+	if err != nil {
+		return err
+	}
+
+	id := d.Id()
+
+	job, err := resourceDataflowJobGetJob(config, project, region, id)
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Dataflow job %s", id))
+	}
+
+	d.Set("state", job.CurrentState)
+	d.Set("name", job.Name)
+	d.Set("project", project)
+	d.Set("labels", job.Labels)
+
+	if _, ok := dataflowTerminalStatesMap[job.CurrentState]; ok {
+		log.Printf("[DEBUG] Removing resource '%s' because it is in state %s.\n", job.Name, job.CurrentState)
+		d.SetId("")
+		return nil
+	}
+	d.SetId(job.Id)
+
+	return nil
+}
+
+// resourceDataflowFlexTemplateJobUpdateByReplacement updates a Flex Template job by replacement.
+func resourceDataflowFlexTemplateJobUpdateByReplacement(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region, err := getRegion(d, config)
+	if err != nil {
+		return err
+	}
+
+	request := buildLaunchFlexTemplateRequest(d, config)
+
+	var job *dataflow.Job
+	err = retryTimeDuration(func() (updateErr error) {
+		job, updateErr = resourceDataflowFlexTemplateJobLaunchTemplate(config, project, region, &request)
+		return updateErr
+	}, time.Minute*time.Duration(5), isDataflowJobUpdateRetryableError)
+	if err != nil {
+		return err
+	}
+	d.SetId(job.Id)
+	return resourceDataflowJobRead(d, meta)
+}
+
+// resourceDataflowFlexTemplateJobLaunchTemplate Launches a Flex Template Job.
```

Comment:
I'd slightly prefer that we inline both methods below. While it's not DRY, it matches the convention in the rest of the provider. If you feel strongly otherwise, feel free to ignore this suggestion.